STM32 SPI: Extend data size support

gautierg-st opened the pull request:
Until now the STM32 SPI driver only supported 8 and 16-bit data sizes, even though some SoCs supports more sizes than that.

This PR extends the possible data sizes that can be used, ranging from 4 to 32-bit depending on SoC and SPI instances.

gautierg-st committed:
dts: bindings: spi: add st,stm32-data-width property

Adds a new property for STM32 SPI bindings to indicate which data widths
are supported by each SPI instance.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: spi: stm32: extend data size support

Extend the number of SPI data sizes supported for series that allow it.
The driver uses the new property st,stm32-data-width to create a table
with all compatible data widths and to check if the SPI instance
support the required width (some instances may have a limited set
of widths compared to others on the same SoC).

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
dts: arm: st: add st,stm32-data-width property for all spi instances

Addd st,stm32-data-width property for all STM32 SPI instances.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: spi: stm32: fix spi rtio with fifo-enable

This commit fixes the SPI RTIO transfers when the fifo-enable property is
enabled.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: spi: loopback: change duration scaling

Increase the duration scaling for two boards that were causing some tests
to fail.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

erwango commented on dts/bindings/spi/st,stm32h7-spi.yaml:
```diff
@@ -32,3 +32,9 @@ properties:
+  limited-instance:
+    type: boolean
+    description: |
+      Indicates if the SPI instance is a limited-features instance.
```
Does it comes with other limitations ?
Otherwise I think we should somehow explicit that this is a data width limitation.

erwango commented on drivers/spi/spi_ll_stm32.h:
```diff
@@ -97,13 +98,17 @@ static inline uint32_t ll_func_dma_get_reg_addr(SPI_TypeDef *spi, uint32_t locat
-		!LL_SPI_IsActiveFlag_BSY(spi));
-#endif /* LL_SPI_SR_TXC */
+	return (!LL_SPI_IsActiveFlag_TXE(spi) ||
+		LL_SPI_IsActiveFlag_BSY(spi));
```
Dedicated commit required ?

mathieuchopstm commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -67,6 +67,55 @@ LOG_MODULE_REGISTER(spi_ll_stm32);
+	_CONCAT_2(LL_SPI_DATAWIDTH_, UTIL_INC(UTIL_INC(UTIL_INC(UTIL_INC(i)))), BIT)
+static const uint32_t table_datawidth[] = {
+	LISTIFY(UTIL_DEC(UTIL_DEC(UTIL_DEC(STM32_SPI_DATAWIDTH_MAX))), STM32_SPI_DATAWIDTH, (,))
+};
```
To be checked if this builds but a more readable implementation could be:
```c
#define N_PLUS_3(n, _) UTIL_INC(UTIL_INC(UTIL_INC(n))))
/* generate a list from 4 to max_width: "4, 5, ..., max_width" */
#define DATA_WIDTH_SIZES(max_width) \
    LISTIFY(UTIL_DEC(UTIL_DEC(UTIL_DEC(max_width))), N_PLUS_3, (,))

#define STM32_SPI_DATAWIDTH(n) CONCAT(LL_SPI_DATAWIDTH_, n)
static const uint32_t table_datawidth[] = {
    FOR_EACH(STM32_SPI_DATAWIDTH, (,), DATA_WIDTH_SIZES(STM32_SPI_DATAWIDTH_MAX))
};
```

gautierg-st committed:
drivers: spi: stm32: invert ll_func_spi_dma_busy logic

Inverts ll_func_spi_dma_busy logic so that the function returns true if
the SPI DMA is busy (as suggested by the name) instead of the other way
around.
Also completes the check for H7-compatible with the LL_SPI_IsActiveFlag_EOT
function when transfer size is set.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

erwango commented on dts/bindings/spi/st,stm32-spi-common.yaml:
```diff
@@ -21,3 +21,16 @@ properties:
+      Indicates the data width supported by the SPI instance.
+      - full-32-bit: supports data width from 4 to 32 bits
+      - full-16-bit: supports data width from 4 to 16 bits
+      - limited-16-bit: supports only 8 and 16 bits data width
```
I'd like to bikeshed on bit on this. What about more explicits:
```
      - full-4-to-32-bit: supports data width from 4 to 32 bits
      - full-4-to-16-bit: supports data width from 4 to 16 bits
      - limited-8-16-bit: supports only 8 and 16 bits data width
```
or more simple:
```
      - 4-to-32-bit: supports data width from 4 to 32 bits
      - 4-to-16-bit: supports data width from 4 to 16 bits
      - 8-16-bit: supports only 8 and 16 bits data width
```
(also not sure if we don't need `s` to `bit` )

gautierg-st replied:
The full/limited terms are explicitly used in the RefMan of H7-compatible SPI instances. So that way, it is immediately identifiable, and they imply some limitations for some instances.
On the other hand, these terms are not used for other series. For each given series, all instances are identical, so there is no need in the RefMan to include those terms. But I think they can be extrapolated for these series (though it doesn't come with all the features of the H7-compatible).
All in all, I'm rather in favor of your first proposition (bonus point: all `:` are nicely aligned :D )

> (also not sure if we don't need `s` to `bit` )

AFAIK, we talk about 32-bit words (32-bit is an adjective, so no `s`), or words of 32 bits (here bits is a noun so it takes an `s`). So I'd say both are fine, it depends on your perspective.

erwango reviewed (changes requested):
Best part of the review remains :-D
Otherwise LGTM

mathieuchopstm commented on drivers/spi/spi_stm32.c:
```diff
@@ -302,43 +378,55 @@ static int spi_dma_move_buffers(const struct device *dev, size_t len)
 		LL_SPI_TransmitData16(spi, tx_frame);
-		spi_context_update_tx(&data->ctx, 2, 1);
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi)
+	} else {
```
Nit: SonarQube would prefer (and I think this looks better too)
```suggestion
	} else {
#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi)
```

mathieuchopstm commented on drivers/spi/spi_stm32.c:
```diff
@@ -410,35 +498,44 @@ static void spi_stm32_shift_m(const struct spi_stm32_config *cfg,
 			UNALIGNED_PUT(rx_frame, (uint16_t *)data->ctx.rx_buf);
-			spi_context_update_rx(&data->ctx, 2, 1);
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi)
+		} else {
```
Ditto:
```suggestion
		} else {
#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi)
```

mathieuchopstm reviewed (approved):
LGTM. A few nits:
